marlowe-integration-tests/test/Language/Marlowe/Runtime/Integration/MarloweQuery.hs

```diff
@@ -0,0 +1,458 @@
+  for_ (Unknown : ((Known . Left <$> allContractNos) <> (Known . Right <$> allTxNos))) \txNo -> do
+    it (show txNo) $ runMarloweQueryIntegrationTest \testData -> do
+      let
+        expected = case txNo of
+          Unknown -> Nothing
+          Known (Left _) -> Nothing
+          Known (Right txNo') -> Just $ txNoToSomeTransaction testData txNo'
+      actual <- getTransaction $ contractOrTxNoToTxId testData txNo
+      liftIO $ actual `shouldBe` expected
+
+getWithdrawalsSpec :: SpecWith MarloweQueryTestData
+getWithdrawalsSpec = describe "getWithdrawals" do
+  for_ (Unknown : (Known <$> allContractNos)) \contractNo -> do
+    it (show contractNo) $ runMarloweQueryIntegrationTest \testData -> do
+      let
+        expected = case contractNo of
+          Unknown -> Nothing
+          Known contractNo' -> Just $ contractNoToWithdrawals testData contractNo'
+      actual <- getWithdrawals $ contractNoToContractId testData contractNo
+      liftIO $ actual `shouldBe` expected
+
+getWithdrawalSpec :: SpecWith MarloweQueryTestData
+getWithdrawalSpec = describe "getWithdrawal" do
+  it "Returns Nothing for a fake txId" $ runMarloweQueryIntegrationTest \_ -> do
+    actual <- getWithdrawal "0000000000000000000000000000000000000000000000000000000000000000"
+    liftIO $ actual `shouldBe` Nothing
+  it "Returns Nothing for a create txId" $ runMarloweQueryIntegrationTest \testData -> do
+    actual <- getWithdrawal $ contractOrTxNoToTxId testData $ Known $ Left Contract1
+    liftIO $ actual `shouldBe` Nothing
+  it "Returns Nothing for an applyInputs txId" $ runMarloweQueryIntegrationTest \testData -> do
+    actual <- getWithdrawal $ contractOrTxNoToTxId testData $ Known $ Right Contract1Step4
+    liftIO $ actual `shouldBe` Nothing
+  it "Returns Just for a withdrawal txId" $ runMarloweQueryIntegrationTest \testData@MarloweQueryTestData{..} -> do
+    actual <- getWithdrawal $ fromCardanoTxId $ getTxId $ fst contract1Step5
+    liftIO $ actual `shouldBe` Just (contract1Step5Withdrawal testData)
+
+setup :: ActionWith MarloweQueryTestData -> IO ()
+setup runSpec = withLocalMarloweRuntime $ runIntegrationTest do
+  runtime <- ask
+  wallet1 <- getGenesisWallet 0
+  wallet2 <- getGenesisWallet 1
+  contract1 <- createStandardContract wallet1 wallet2
+  contract1Step1 <- makeInitialDeposit contract1
+  contract1Step2 <- chooseGimmeTheMoney contract1Step1
+  contract2 <- createStandardContract wallet1 wallet2
+  contract1Step3 <- sendNotify contract1Step2
+  contract1Step4 <- makeReturnDeposit contract1Step3
+  contract1Step5 <- withdrawPartyAFunds contract1Step4
+  contract2Step1 <- makeInitialDeposit contract2
+  contract3 <- createStandardContract wallet1 wallet2
+  contract4 <- createStandardContract wallet1 wallet2
+  contract2Step2 <- chooseGimmeTheMoney contract2Step1
+  contract2Step3 <- sendNotify contract2Step2
+  contract2Step4 <- makeReturnDeposit contract2Step3
+  contract3Step1 <- makeInitialDeposit contract3
```

Comment:
Could we add two more types of steps to one or more of the contracts?
- multiple inputs are applied in the same transaction
- empty input is applied because the contract has timed out.